File: 01-factory-method/01-factory-method.php
<?php

// Product Interface
interface Notification
{
    public function send(string $message): void;
}

// Concrete Class
class EmailNotification implements Notification
{
    public function send(string $message): void
    {
        echo "Sent Email: $message";
    }
}

// Concrete Class
class SmsNotification implements Notification
{
    public function send(string $message): void
    {
        echo "Sent SMS: $message";
    }
}

// Factory Method in the Base Class
abstract class NotificationService
{
    abstract protected function createNotification(): Notification;

    public function notify(string $message): void
    {
        $notification = $this->createNotification();
        $notification->send($message);
    }
}

// Concrete Factory
class EmailService extends NotificationService
{
    protected function createNotification(): Notification
    {
        return new EmailNotification();
    }
}

// Concrete Factory
class SmsService extends NotificationService
{
    protected function createNotification(): Notification
    {
        return new SmsNotification();
    }
}

// Client Code
class Client
{
    public function __construct(protected NotificationService $service)
    {
    }

    public function sendMessage(string $message = "Default Message"): void
    {
        $this->service->notify($message);
    }
}

// Usage
$client = new Client(new EmailService());
$client->sendMessage("Hello via Email!"); // Outputs: Sent Email: Hello via Email!

$client = new Client(new SmsService());
$client->sendMessage("Hello via SMS!"); // Outputs: Sent SMS: Hello via SMS!

File: 01-factory-method/01-factory-method.go
package main

import "fmt"

// PaymentGateway is a product interface
type PaymentGateway interface {
	pay(n int)
}

// Stripe is a concrete product
type Stripe struct {
	Amount int
}

func (s Stripe) pay(n int) {
	fmt.Print("paid by stripe: ", n)
}

// Paypal is a concrete product
type Paypal struct {
	Amount int
}

func (p Paypal) pay(n int) {
	fmt.Print("paid by paypal: ", n)
}

type PaymentFactory interface {
	createPayment(amount int) PaymentGateway
}

type StripeFactory struct{}

func (s StripeFactory) createPayment(amount int) PaymentGateway {
	return newStripe(amount)
}

type PaypalFactory struct{}

func (p PaypalFactory) createPayment(amount int) PaymentGateway {
	return newPaypal(amount)
}

// Factory method
func newStripe(n int) *Stripe {
	return &Stripe{
		Amount: n,
	}
}

// Factory method
func newPaypal(n int) *Paypal {
	return &Paypal{
		Amount: n,
	}
}

// Client code
func processPayment(factory PaymentFactory, amount int) {
	payment := factory.createPayment(amount)
	payment.pay(amount)
}

func main() {
	processPayment(StripeFactory{}, 100) // Outputs: paid by stripe 100
	processPayment(PaypalFactory{}, 100) // Outputs: paid by paypal 100
}
